### PR TITLE
WIP Feature/refactor views

### PR DESCRIPTION
**Pull request by SiggyF**
Restructure code to make the selection logic a bit more robust.
# Functionality
- [x] Combined scenario calculations

# Bugs
- [x] Crashing browser after selecting combine scenario's (due to circular events in storewatch)
- [x] NULL notifications 
- [x] Opacity for marker layers

# Clean code
- [x] make naming more consistent (layerSet|layerset|map), (feature|breach|scenario), (layer|variant), (panel), (active|selected|visible) (LayersetList|LayerSetList)
- [x] comments and more comments
- [x] cleanup normalize, layer cleaning and notifications
- [x] remove long list of lodash imports 
- [x] remove unused code
- [x] get rid of unused code in store

# Maintainability
- [x] update vue so we can use [named slots](https://vuejs.org/v2/guide/components-slots.html#Named-Slots) to add buttons to the layer panel
- [x] move code custom buttons to combine view and add them in the slot
- [x] add custom icon slot to layerPanel
- [x] move page title setting code to 1 place (App.vue)
- [x] reuse css across views

# Keep things in the right pace
- [x] Separate view for Combine/Select and rest of the maps
- [x] Move scenario selection logic from store to Combine vue

- [x] remove breach logic from map
- [x] Use route properties to pass properties behaviour
- [x] pass layer as property for objects that need present the layer (LayerControl, LayerControlList, etc..)
- [x] avoid sub/sub component store interaction, use [.sync](https://vuejs.org/v2/guide/components-custom-events.html#sync-Modifier) where appropriate
- [x] remove double administration (collapsed, activeIds, selectedIds)
- [x] avoid double administration of ids and objects (selectedFeature, selectedLayer, selectedVariant)

# Minor ui fixes
- [x] trigger lifecycle hooks of components across different routes, using a [key](https://vuejs.org/v2/api/#key)
- [x] replace opacity select with slider
- [x] put filter interface in popup, as filter was applied across all layers
- [x] use a bit more space
- [x] format filter menu

# Build environment
- [x] put base url in [dotenv](https://cli.vuejs.org/guide/mode-and-env.html#environment-variables)  file 
- [x] add new e2e tests
- [x] reduce lodash file size

# Left open
- [ ] Leaflet-directive is not consistent with [vue documentation](https://vuejs.org/v2/guide/components-edge-cases.html#Accessing-the-Parent-Component-Instance) and [cookbook](https://vuejs.org/v2/cookbook/practical-use-of-scoped-slots.html#ad).  It is also flickering and slow. 
- [ ] layers are not used consistently, there are now three different schema's (layers as loaded by a layerSet, flattened layers (without .map) and Leaflet layers.

### CODE REVIEW

File: src/lib/leaflet-directive.js [hunk 1]
```diff
@@ -4,7 +4,9 @@ import L from '@/lib/leaflet-utils/leaf'
 
 let map
 let layerGroup
-
+// TODO: replace  this with vue2-leaflet
```
Comment (SiggyF) [inline src/lib/leaflet-directive.js]: This leaflet-directive is a bad idea:
- not inspectable
- no hierarchy 
- not reusable

I believe the idea was to avoid a double administration of doms. Leaflet is responsible for building the dom elements in it's own component.  If you have a custom element (video, canvas, marker), then you can let vue take over.  You can always use Leaflet directly without a vue component by accessing the [mapObject](https://korigan.github.io/Vue2Leaflet/#/quickstart?id=accessing-leaflet-api)

File: src/lib/leaflet-utils/layer-factory.js [hunk 2]
```diff
@@ -4,42 +4,33 @@ import 'leaflet.markercluster'
 import 'leaflet.markercluster/dist/MarkerCluster.css'
 import 'leaflet.markercluster/dist/MarkerCluster.Default.css'
 
-import {
```
Comment (SiggyF) [inline src/lib/leaflet-utils/layer-factory.js]: moved to cleanLayers for now

File: src/lib/layer-parser.js [hunk 3]
```diff
@@ -1,23 +1,115 @@
...
+} from '@/lib/liwo-identifiers'
+import {
+  blackIcon,
+  greenIcon,
+  redIcon
+} from '@/lib/leaflet-utils/markers'
+
+const DEFAULT_ICON = new L.Icon.Default()
+
+export function flattenLayerSet (layerSet) {
```
Comment (SiggyF) [inline src/lib/layer-parser.js]: This generates the layers as I would like to see them.

File: src/lib/build-layerset-notifications.js [hunk 4]
```diff
@@ -1,27 +1,87 @@
...
-    .map(({ map, notify }) => ({ id: map.layer, notification: notify }))
-}
+import _ from 'lodash'
 
-function mapLayers (layers, layerId) {
-  return layers
-    .map(({ id, notify, variants: _variants }) => {
-      const variants = mapVariants(_variants) || []
-      return { id: layerId, notification: notify, variants }
+export default function buildLayerSetNotifications (layers) {
```
Comment (SiggyF) [inline src/lib/build-layerset-notifications.js]: this now searches for notifications in both layers and features.

File: src/components/NotificationBar.vue [hunk 5]
```diff
@@ -44,22 +53,38 @@
...
-      validator (value) {
-        return typeof value === 'string'
-      }
+    notifications: {
+      type: Array
+    }
+  },
+  data () {
+    return {
+      editableNotifications: []
```
Comment (SiggyF) [inline src/components/NotificationBar.vue]: props are not editable. Here we want to change the show state.

File: src/components/NotificationBar.vue [hunk 6]
```diff
@@ -44,22 +53,38 @@
 </style>
 
 <script>
+import _ from 'lodash'
 export default {
   props: {
-    type: {
```
Comment (SiggyF) [inline src/components/NotificationBar.vue]: all notifications

File: src/components/LiwoMap.vue [hunk 7]
```diff
@@ -108,7 +90,7 @@ export default {
     width: calc(100% - 2rem);
     display: block;
     margin: 0 auto;
-    height: calc(100vh - 20rem);
```
Comment (SiggyF) [inline src/components/LiwoMap.vue]: Get rid of unfilled lower grey area

File: src/components/NotificationBar.vue [hunk 8]
```diff
@@ -1,11 +1,20 @@
 <template>
-  <aside class="notification-bar"
```
Comment (SiggyF) [inline src/components/NotificationBar.vue]: Now contains all notifications

File: src/components/LiwoMap.vue [hunk 9]
```diff
@@ -82,23 +81,6 @@ export default {
...
-  watch: {
-    parsedLayerSet (parsedLayerSet) {
-      if (!parsedLayerSet) {
-        return []
-      }
-
-      parsedLayerSet
-        .then(
-          (layers) => {
-            this.expandedMapLayers = Object.freeze(layers)
```
Comment (SiggyF) [inline src/components/LiwoMap.vue]: Use properties for unmutable objects.

File: src/components/LiwoMap.vue [hunk 10]
```diff
@@ -43,10 +43,9 @@ export default {
       'layerUnits',
       'selectedLayerId',
       'visibleVariantIndexByLayerId',
-      'activeLayerSetId'
```
Comment (SiggyF) [inline src/components/LiwoMap.vue]: activeLayerSetId and mapId were the same, now layerSetId

File: src/components/LiwoMap.vue [hunk 11]
```diff
@@ -15,11 +15,12 @@
 import { mapState, mapGetters } from 'vuex'
 
 import createMapConfig from '@/lib/leaflet-utils/mapconfig-factory'
-import buildBreachNotifications from '@/lib/build-breach-notifications'
```
Comment (SiggyF) [inline src/components/LiwoMap.vue]: breachnotifications treated the same as other notifications

File: src/components/LayerPanelItem.vue [hunk 12]
```diff
@@ -71,7 +64,6 @@ export default {
   },
   methods: {
     setActiveLayer () {
-      this.$store.commit('setActiveLayerSetId', this.layerId)
```
Comment (SiggyF) [inline src/components/LayerPanelItem.vue]: same as mapId

File: src/components/LayerPanelItem.vue [hunk 13]
```diff
@@ -71,7 +64,6 @@ export default {
   },
   methods: {
     setActiveLayer () {
```
Comment (SiggyF) [inline src/components/LayerPanelItem.vue]: ActiveLayer or ActiveLayerSet

File: src/components/LayerPanelItem.vue [hunk 14]
```diff
@@ -50,15 +50,8 @@ export default {
     collapsed: false
   }),
   computed: {
-    ...mapState([
-      'activeLayerSetId'
-    ]),
     layerControlListIsVisible () {
-      if (!this.activeLayerSetId || !this.layerId) {
```
Comment (SiggyF) [inline src/components/LayerPanelItem.vue]: unclear what this does

File: src/components/LayerPanel.vue [hunk 15]
```diff
@@ -23,42 +23,9 @@
...
-          @click="$emit('open-export')"
         >
-          <svg class="icon" xmlns="http://www.w3.org/2000/svg" width="32" height="32" viewBox="0 0 24 24">
-            <path fill="none" d="M0 0h24v24H0z"/>
-            <path d="M18 17v2H6v-2H3v4c0 .6.4 1 1 1h16c.6 0 1-.4 1-1v-4h-3z"/>
-            <path d="M11 16.5a1.4 1.4 0 0 0 2 0l5.8-7.3a1.4 1.4 0 0 0-1.7-2l-3.1 2V3.4c0-1-1-1.4-2-1.4s-2 .3-2 1.4v5.8l-3-2a1.4 1.4 0 0 0-1.8 2l5.7 7.3z"/>
-          </svg>
-          Kaart exporteren
-        </button>
+        <slot></slot>
```
Comment (SiggyF) [inline src/components/LayerPanel.vue]: allows to insert content

File: src/components/LayerPanel.vue [hunk 16]
```diff
@@ -23,42 +23,9 @@
       </div>
       <footer
         class="layer-panel__actions"
-      >
```
Comment (SiggyF) [inline src/components/LayerPanel.vue]: moved to combine view

File: src/components/LiwoMap.vue [hunk 17]
```diff
@@ -5,7 +5,7 @@
     v-leaflet="{
       callbacks: { breachCallBack, initMapObject },
       config: mapConfig,
-      mapLayers: [ ...expandedMapLayers ].reverse(),
```
Comment (SiggyF) [inline src/components/LiwoMap.vue]: expansion is done while loading

File: src/lib/layer-parser.js [hunk 18]
```diff
@@ -1,23 +1,115 @@
...
+  // now we have layers[variants[]]
+  // so we just need to flatten
+  return _.flatten(layers)
+}
+
 export function normalizeLayer (layer) {
-  return {
+  let firstVariant = _.first(layer.variants)
+  // namespace should be available to legend
+  let namespace = _.get(firstVariant, 'map.namespace', '')
```
Comment (SiggyF) [inline src/lib/layer-parser.js]: what is a namespace in this context?

File: src/lib/load-layersets.js [hunk 19]
```diff
@@ -1,41 +1,97 @@
...
+  return layerSet
 }
 
-export function loadLayersets () {
-  const body = JSON.stringify({ username: 'marko@voorhoede.nl', password: 'DikkeDoei123', mode: '' })
+export function loadLayerSets () {
+  // we don't need to login anymore, but the function is still used to
+  // get the list of public maps
+  const body = JSON.stringify({
+    username: 'marko@voorhoede.nl',
```
Comment (SiggyF) [inline src/lib/load-layersets.js]: remove this

File: src/views/Combine.vue [hunk 20]
```diff
@@ -0,0 +1,281 @@
+<template>
```
Comment (SiggyF) [inline src/views/Combine.vue]: this is rather different from the other views. This view:
- knows about breaches
- provides custom interactivity
- loads extra layers
- has different buttons

File: src/views/Viewer.vue [hunk 21]
```diff
@@ -1,152 +1,98 @@
...
+
+    </layer-panel>
+    <legend-panel
+      v-if="visibleLayerLegend"
+      :caption="visibleLayerLegend.title"
+      :namespace="visibleLayerLegend.namespace"
+      :layer-name="visibleLayerLegend.layer"
+      :style-name="visibleLayerLegend.style"
       />
-      <legend-panel
```
Comment (SiggyF) [inline src/views/Viewer.vue]: remove the specific breach code

File: vue.config.js [hunk 22]
```diff
@@ -3,7 +3,7 @@ const CopyWebpackPlugin = require('copy-webpack-plugin')
 
 module.exports = {
   // transpileDependencies: [ 'delay' ],
-  baseUrl: process.env.BASE_URL || '/',
```
Comment (SiggyF) [inline vue.config.js]: renamed by vue
